module/zfs/vdev_indirect.c

```diff
@@ -1505,74 +1518,67 @@ vdev_indirect_reconstruct_io_done(zio_t *zio)
 		attempts_max = zfs_reconstruct_indirect_combinations_max;
 
 	/*
-	 * Discard duplicate copies of split segments to minimize the
-	 * number of unique combinations when attempting reconstruction.
+	 * Determine the unique children for a split segment and add them
+	 * to the is_unique_child list.  By restricting reconstruction
+	 * attempts to these children, only unique combinations will be
+	 * considered.  This can vastly reduce the search space when there
+	 * are a large number of indirect splits.
 	 */
 	for (indirect_split_t *is = list_head(&iv->iv_splits);
 	    is != NULL; is = list_next(&iv->iv_splits, is)) {
-		uint64_t is_copies = 0;
+		is->is_unique_children = 0;
 
 		for (int i = 0; i < is->is_children; i++) {
-			if (is->is_child[i].ic_data == NULL)
+			indirect_child_t *ic_i = &is->is_child[i];
+
+			if (ic_i->ic_data == NULL ||
+			    ic_i->ic_duplicate != NULL)
 				continue;
 
 			for (int j = i + 1; j < is->is_children; j++) {
-				if (is->is_child[j].ic_data == NULL)
+				indirect_child_t *ic_j = &is->is_child[j];
+
+				if (ic_j->ic_data == NULL ||
+				    ic_j->ic_duplicate != NULL)
 					continue;
 
-				if (is->is_child[j].ic_duplicate == -1 &&
-				    abd_cmp(is->is_child[i].ic_data,
-				    is->is_child[j].ic_data) == 0) {
-					is->is_child[j].ic_duplicate = i;
-				}
+				if (abd_cmp(ic_i->ic_data, ic_j->ic_data) == 0)
+					ic_j->ic_duplicate = ic_i;
 			}
 
-			is_copies++;
+			is->is_unique_children++;
+			list_insert_tail(&is->is_unique_child, ic_i);
 		}
 
-		/* Reconstruction is impossible, no valid is->is_child[] */
-		if (is_copies == 0) {
+		/* Reconstruction is impossible, no valid children */
+		EQUIV(list_is_empty(&is->is_unique_child),
+		    is->is_unique_children == 0);
+		if (list_is_empty(&is->is_unique_child)) {
 			zio->io_error = EIO;
 			vdev_indirect_all_checksum_errors(zio);
 			zio_checksum_verified(zio);
 			return;
 		}
 
-		combinations *= is_copies;
+		combinations *= is->is_unique_children;
+		is->is_good_child = list_head(&is->is_unique_child);
 	}
 
 	for (;;) {
```

Comment:
So I remember looking at this code the first time and it does make sense although I found it a bit hard to read due to the fact that we loop back here and then deciding again if we are enumerating or randomly picking (plus thinking about `attempts` and `more`).

I believe it is fine leaving the code as it is, on the other hand I thought of the layout below which may be a bit simpler. Basically in case of having this infinite loop we have the following:
```
int ret;
if (combinations <= attempts_max) {
   ret = vdev_indirect_splits_enumerate_all(iv, zio);
} else {
   ret = vdev_indirect_splits_enumerate_randomly(iv, zio, max_attempts);
}

if (ret != 0) {
    /* All attempted combinations failed. */
    zio->io_error = ret;
    vdev_indirect_all_checksum_errors(zio);
    zio_checksum_verified(zio);
}
```

where:
```
int
vdev_indirect_splits_checksum_validate(iv)
{
    for (indirect_split_t *is = list_head(&iv->iv_splits);
        is != NULL; is = list_next(&iv->iv_splits, is)) {
        ASSERT3P(is->is_good_child->ic_data, !=, NULL);
        ASSERT3P(is->is_good_child->ic_duplicate, ==, NULL);
        abd_copy_off(zio->io_abd, is->is_good_child->ic_data,
            is->is_split_offset, 0, is->is_size);
    }
    ... zbc;
    return(zio_checksum_error(zio, &zbc));
}

int
vdev_indirect_splits_enumerate_all(iv, zio)
{
    ... do the enumeration part ... assert that you don't do more than the *combinations* variable ....
   ... check every combination with vdev_indirect_splits_checksum_validate() ... check its return value ...
}

int
 vdev_indirect_splits_enumerate_randomly(iv, zio, max_attempts)
{
    for (attempt = 0; attempt < max_attempts; attempt++) {
      ... do the spa_get_random_part , and check every time with vdev_indirect_splits_checksum_validate() ...
      ... break when you match .... [you can also have a dprintf saying ("matched after %d attempts", attempt)]
    }
}
```

The disadvantage of the above layout is that we have auxiliary functions but hopefully make the code more readable (instead of having one big loop that decides if we are randomly enumerating over some stuff or enumerating through everything, in every iteration, we have two separate loops with more straightforward logic). An additional benefit may be that we can more easily trace things with DTrace/eBPF because of the auxiliary functions, and gives us more introspection about which strategy we decide).

Again, if you think it doesn't worth it or there is something wrong feel free to drop this.
The code LGTM as it is.